### PR TITLE
[expression.dd] Improve IndexExpression & SliceExpression docs

### PR DESCRIPTION
**Pull request by ntrel**
* Make IndexExpression docs more generic.
* Mention pointer indexing & slicing.

* `$` is also declared for *ValueSeq* slices and can be for user types - make this paragraph independent of array slicing.
* Don't only talk about slicing arrays throughout, but slicing elements of the *PostfixExpression* instead. This is more generic.
* Mention first and last element expressions for arrays and pointers.
* Add *Slice Conversion to Static Array* subheading, move *ValueSeq* paragraph above it & add a *ValueSeq* link.

### CODE REVIEW

File: spec/expression.dd [hunk 1]
```diff
@@ -1266,43 +1266,51 @@ $(GNAME Slice):
...
+        If $(I PostfixExpression) is an expression of static array or
+        dynamic array type, the result of the slice is a dynamic array
+        of the element type of the $(I PostfixExpression).
+    )
+
+    $(P If $(I PostfixExpression) is a $(DDSUBLINK spec/template, variadic-templates, $(I ValueSeq)), then
+        the result of the slice is a new $(I ValueSeq) formed
+        from the upper and lower bounds, which must statically evaluate
+        to integral constants.
+        It is an error if those bounds are out of range.
```
Comment (maxhaton) [inline spec/expression.dd:1278]:
```suggestion
        These bounds must be in range.
```

Comment (maxhaton) [reply spec/expression.dd:1278]: Reads a little bit nicer + not a huge fan of language specs mentioning "error" and similar.

File: spec/expression.dd [hunk 2]
```diff
@@ -1266,43 +1266,51 @@ $(GNAME Slice):
...
+    $(P If the $(D [ ]) form is used, the slice is of all the elements in $(I PostfixExpression).
     )
 
     $(P A $(I SliceExpression) is not a modifiable lvalue.)
 
+$(H3 $(LNAME2 slice_to_static_array, Slice Conversion to Static Array))
+
     $(P If the slice bounds can be known at compile time, the slice expression
-    is implicitly convertible to an lvalue of static array. For example:)
+    may be implicitly convertible to an lvalue of static array. For example:)
```
Comment (dkorpel) [inline spec/expression.dd:1303]: Why `may be`, are there cases where it isn't implicitly convertible?

Comment (ntrel) [reply spec/expression.dd:1318]: @dkorpel The *PostfixExpression* might be a user-defined type where `opSlice` doesn't return a dynamic array.

**Comment by ntrel** [comment]
This pull now covers `IndexExpression` as well as it's closely related. I also updated the first paragraph of the `SliceExpression` docs and added a paragraph about slicing a pointer.

File: spec/expression.dd (hunk 1 above)
Comment (ntrel) [reply spec/expression.dd:1293]: 'must be' is not as clear that this is checked at compile time. (I didn't change the original sentence, just moved it).

File: spec/expression.dd [hunk 3]
```diff
@@ -1235,22 +1235,29 @@ $(GNAME IndexExpression):
...
-$(GLINK ArgumentList) and $(DOLLAR) appears in that scope only.)
-
-    $(P If $(I PostfixExpression) is a $(I ValueSeq),
-        then the $(GLINK ArgumentList) must consist of only one argument,
+    $(P $(I PostfixExpression) is evaluated.
+        If $(I PostfixExpression) is an expression of static or
+        dynamic array type, the result of the indexing is an lvalue
+        of the *i*th element in the array.
+        If $(I PostfixExpression) is a pointer `p`, the result is `*(p + i)`.
+        In each case, `i` is an integer evaluated from $(I ArgumentList).
```
Comment (dkorpel) [inline spec/expression.dd:1243]: You use `i` twice and only define what it refers to afterwards. I think that's a bit awkward, I'd rather see `i` defined first.

Comment (ntrel) [reply spec/expression.dd]: I have now defined `i` in the same sentence as the first use, hopefully that reads better.

**Review by dkorpel — APPROVED** [review]

**Pull request merged**